src/audio/sound_dose/sound_dose.c

```diff
@@ -0,0 +1,376 @@
+	cd->feature = (struct sof_audio_feature *)cd->abi->data;
+	cd->feature->data_size = sizeof(struct sof_sound_dose);
+	cd->feature->type = SOF_AUDIO_FEATURE_SOUND_DOSE_MEL;
+	cd->feature->num_audio_features = 1; /* Single MEL value in audio feature data */
+	cd->dose = (struct sof_sound_dose *)cd->feature->data;
+	return 0;
+}
+
+/**
+ * sound_dose_init() - Initialize the component.
+ * @mod: Pointer to module data.
+ *
+ * This function is called when the instance is created. The
+ * macro __cold informs that the code that is non-critical
+ * is loaded to slower but large DRAM.
+ *
+ * Return: Zero if success, otherwise error code.
+ */
+__cold static int sound_dose_init(struct processing_module *mod)
+{
+	struct module_data *md = &mod->priv;
+	struct comp_dev *dev = mod->dev;
+	struct sound_dose_comp_data *cd;
+	int ret;
+
+	comp_info(dev, "sound_dose_init()");
+
+	cd = rzalloc(SOF_MEM_FLAG_USER, sizeof(*cd));
+	if (!cd)
+		return -ENOMEM;
+
+	md->private = cd;
+
+	sound_dose_setup_init(cd);
+	ret = sound_dose_audio_feature_init(mod);
+	if (ret) {
+		rfree(cd);
+		return ret;
+	}
+
+	ret = sound_dose_ipc_notification_init(mod);
+	if (ret) {
+		rfree(cd->abi);
+		rfree(cd);
+	}
+
+	return ret;
+}
+
+/**
+ * sound_dose_process() - The audio data processing function.
+ * @mod: Pointer to module data.
+ * @sources: Pointer to audio samples data sources array.
+ * @num_of_sources: Number of sources in the array.
+ * @sinks: Pointer to audio samples data sinks array.
+ * @num_of_sinks: Number of sinks in the array.
+ *
+ * This is the processing function that is called for scheduled
+ * pipelines. The processing is controlled by the enable switch.
+ *
+ * Return: Zero if success, otherwise error code.
+ */
+static int sound_dose_process(struct processing_module *mod,
+			      struct sof_source **sources,
+			      int num_of_sources,
+			      struct sof_sink **sinks,
+			      int num_of_sinks)
+{
+	struct sound_dose_comp_data *cd = module_get_private_data(mod);
+	struct comp_dev *dev = mod->dev;
+	struct sof_source *source = sources[0]; /* One input in this example */
+	struct sof_sink *sink = sinks[0]; /* One output in this example */
+	int32_t arg;
+	int frames = source_get_data_frames_available(source);
+	int sink_frames = sink_get_free_frames(sink);
+
+	comp_dbg(dev, "sound_dose_process()");
+
+	if (cd->gain_update) {
+		/* Convert dB * 100 to Q8.24 */
+		arg = (int32_t)cd->att.gain * SOUND_DOSE_ONE_OVER_100_Q24;
+
+		/* Calculate linear value as Q12.20 and convert to Q2.30 */
+		cd->new_gain = Q_SHIFT_LEFT(sofm_db2lin_fixed(arg), 20, 30);
+		cd->gain_update = false;
+	}
+
+	if (cd->new_gain < cd->gain) {
+		cd->gain = Q_MULTSR_32X32((int64_t)cd->gain, SOUND_DOSE_GAIN_DOWN_Q30, 30, 30, 30);
+		cd->gain = MAX(cd->gain, cd->new_gain);
+	} else if (cd->new_gain > cd->gain) {
+		cd->gain = Q_MULTSR_32X32((int64_t)cd->gain, SOUND_DOSE_GAIN_UP_Q30, 30, 30, 30);
+		cd->gain = MIN(cd->gain, SOUND_DOSE_GAIN_ONE_Q30);
+	}
```

Comment:
should this check be done only when cd->gain_update was true? If so, move it within the if () case?